packages/gatsby/src/utils/api-runner-node.js

```diff
@@ -220,14 +221,36 @@ const runAPI = (plugin, api, args) => {
     } else {
       const result = gatsbyNode[api](...apiCallArgs)
       pluginSpan.finish()
-      return Promise.resolve(result).then(res => {
-        apiFinished = true
-        return res
-      })
+      chain = Promise.resolve(result)
+        .then(res => {
+          /*
+           * validatePluginOptions is special
+           * it validates plugin options and (optionally) mutates them
+           */
+          if (api === `validatePluginOptions`) {
+            if (res) {
+              let resChain = Promise.resolve(res)
+              if (res.validate) {
+                resChain = res.validate(plugin.pluginOptions, {
+                  abortEarly: false,
+                  allowUnknown: true,
+                })
+              }
+              return resChain.then(options => {
```

Comment:
Here's where options are actually coerced. A "gotcha" here is that whether using Joi or not, the returned structure is what will be passed to other APIs that consume plugin options.